File: scripts/things/things_now.py
#!/usr/bin/env python3
"""Arrange selected tasks using an ID-bound Now tag, preserving their projects."""
import argparse
import shutil
import fcntl
import json
import os
from pathlib import Path
import subprocess
import tempfile

ROOT = Path(os.environ.get('THINGS_NOW_DIR', Path.home() / '.local/share/things-now'))

def save(state):
    fd, name = tempfile.mkstemp(prefix='.state-', dir=ROOT)
    with os.fdopen(fd, 'w') as file:
        json.dump(state, file, ensure_ascii=False)
        file.flush()
        os.fsync(file.fileno())
    os.replace(name, ROOT / 'state.json')

def main():
    parser=argparse.ArgumentParser()
    parser.add_argument('action',choices=['ensure','enqueue','show','migrate'])
    parser.add_argument('--tasks',default='[]')
    args=parser.parse_args()
    tasks=json.loads(args.tasks)
    if not isinstance(tasks,list) or any(not isinstance(t.get('id'),str) or not t['id'] for t in tasks):
        raise ValueError('Expected selected task IDs')
    ROOT.mkdir(parents=True,exist_ok=True,mode=0o700)
    with (ROOT/'state.lock').open('a') as lock:
        try: fcntl.flock(lock,fcntl.LOCK_EX|fcntl.LOCK_NB)
        except BlockingIOError: return
        path=ROOT/'state.json'
        state=json.loads(path.read_text()) if path.exists() else {'schema':1,'project_id':'','origins':{}}
        if state.get('schema') == 1:
            if path.exists() and not (ROOT/'state.before-tags.json').exists():
                shutil.copy2(path, ROOT/'state.before-tags.json')
            state.update(schema=2, legacy_project_id=state.pop('project_id', ''), tag_id='')
        if state.get('schema') != 2: raise ValueError('Unknown Now state schema')
        if args.action == 'migrate': tasks=list(state.get('origins', {}).values())
        save(state)
        result=subprocess.run(['/usr/bin/osascript','-l','JavaScript',str(Path(__file__).with_suffix('.js')),
                               'ensure' if args.action == 'show' else args.action,state['tag_id'],json.dumps(tasks,ensure_ascii=False),state.get('legacy_project_id','')],
                              capture_output=True,text=True,timeout=30,check=True)
        result=json.loads(result.stdout)
        state['tag_id']=result['tag_id']
        state['last_result']=result
        save(state)
        if args.action == 'show':
            subprocess.run(['/bin/bash',str(Path(__file__).with_name('toggle_things.sh')), 'tag:' + state['tag_id']],check=True)
        print(json.dumps(result,ensure_ascii=False))
        if args.action == 'enqueue':
            message=f"已编排 {len(result['added']) + len(result['already'])} 项 · Now"
            if result['skipped']: message += f" · 跳过 {len(result['skipped'])} 项"
            subprocess.run(['/usr/bin/python3', str(Path(__file__).with_name('things_timer.py')), 'notice', '--message', message], check=False)
        if result['errors']: raise RuntimeError('Some tasks could not be arranged; see last_result in Now state')

if __name__=='__main__':main()

File: scripts/things/things_now.js
function run(argv) {
    var t = Application('com.culturedcode.ThingsMac');
    var action = argv[0], bound = argv[1], tag;
    if (bound) {
        tag = t.tags.byId(bound);
        if (!tag.exists()) throw Error('Bound Now tag is unavailable');
    } else {
        var matches = t.tags.whose({name:'Now'})();
        if (matches.length > 1) throw Error('More than one Now tag');
        if (matches.length) tag = matches[0];
        else { tag = t.Tag({name:'Now'}); t.tags.push(tag); }
    }
    var added = [], already = [], restored = [], skipped = [], errors = [];
    if (action === 'enqueue' || action === 'migrate') {
        JSON.parse(argv[2]).forEach(function(task) {
            try {
                var item = t.toDos.byId(task.id);
                if (!item.exists() || t.projects.byId(task.id).exists() || t.lists.byId('TMTrashListSource').toDos.byId(task.id).exists()) {
                    skipped.push(task.id); return;
                }
                if (action === 'enqueue' && String(item.status()) !== 'open') { skipped.push(task.id); return; }
                if (action === 'migrate') {
                    if (item.project().id() !== argv[3]) { skipped.push(task.id); return; }
                    if (!task.project_id || !t.projects.byId(task.project_id).exists()) throw Error('Original project unavailable; task left in place');
                }
                if (item.tags().some(function(x) { return x.id() === tag.id(); })) already.push(task.id);
                else {
                    var names = item.tagNames();
                    item.tagNames = names ? names + ', ' + tag.name() : tag.name();
                    if (!item.tags().some(function(x) { return x.id() === tag.id(); })) throw Error('Tag did not persist');
                    added.push(task.id);
                }
                if (action === 'migrate') {
                    item.project = t.projects.byId(task.project_id);
                    if (item.project().id() !== task.project_id) throw Error('Restore did not persist');
                    restored.push(task.id);
                }
            } catch (error) { errors.push({id:task.id, message:String(error)}); }
        });
    }
    return JSON.stringify({tag_id:tag.id(), added:added, already:already, restored:restored, skipped:skipped, errors:errors});
}
